sympy/geometry/line.py

```diff
@@ -1083,24 +1093,68 @@ class Line(LinearEntity):
     >>> s = Segment((0, 0), (0, 1))
     >>> Line(s).equation()
     x
+
+    >>> # a circle object is returned
+    >>> from sympy.abc import x, y, a, b
+    >>> Line(3 * x + y + 18)
+    Line2D(Point2D(0, 0), Point2D(1, -3))
+    >>> Line(3 * a + b + 18, x='a', y='b')
+    Line2D(Point2D(0, 0), Point2D(1, -3))
     """
 
-    def __new__(cls, p1, p2=None, **kwargs):
-        if isinstance(p1, LinearEntity):
-            if p2:
-                raise ValueError('If p1 is a LinearEntity, p2 must be None.')
-            dim = len(p1.p1)
-        else:
-            p1 = Point(p1)
-            dim = len(p1)
-            if p2 is not None or isinstance(p2, Point) and p2.ambient_dimension != dim:
-                p2 = Point(p2)
+    def __new__(cls, *args, **kwargs):
+
+        if len(args) == 1 and isinstance(args[0], Expr):
+            x = kwargs.get('x', 'x')
+            y = kwargs.get('y', 'y')
+            equation = args[0]
+
+            def find(x_, equation_):
+                free = equation_.free_symbols
+                xs = [i for i in free if (i.name if type(x_) is str else i) == x_]
+                if not xs:
+                    raise ValueError('could not find %s' % x_)
+                if len(xs) != 1:
+                    raise ValueError('ambiguous %s' % x_)
+                return xs[0]
+
+            x = find(x, equation)
+            y = find(y, equation)
+
+            if (len(solve(equation, y)) == 0) and (len(solve(equation, x)) != 0) and (degree(poly(equation, x)) == 1):
```

Comment:
This if-else can use `linear_coeffs`:

```python
xin, yin = x, y
x = find(x, equation) or Dummy()
y = find(y, equation)  or Dummy()
a, b, c = linear_coeffs(eq, x, y)
if b:
    return Line((0,-c/b), slope=-a/b)
if a:
    return Line((-c/a, 0), slope=oo)
raise ValueError('neither %s nor %s were found in the equation' % (xin, yin))
```

Reply:
@smichr I tried that approach. Unfortunately it fails tests where `equation = y/x` or `equation = x/y`. But I have figured out (what I think is) a better approach. Please take a look.